src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -313,22 +342,20 @@ public virtual IEnumerable<IEdmOperationImport> ResolveOperationImports(IEdmMode
         /// <returns>The resolved key list.</returns>
         public virtual IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IList<string> positionalValues, Func<IEdmTypeReference, string, object> convertFunc)
         {
-            var keyProperties = type.Key().ToList();
-
             // Throw an error if key size from url doesn't match that from model.
             // Other derived ODataUriResolver intended for alternative key resolution, such as the built in AlternateKeysODataUriResolver,
             // should override this ResolveKeys method.
-            if (keyProperties.Count != positionalValues.Count)
+            if (type.Key().Count() != positionalValues.Count)
             {
                 throw ExceptionUtil.CreateBadRequestError(Strings.BadRequest_KeyCountMismatch(type.FullName()));
             }
 
             var keyPairList = new List<KeyValuePair<string, object>>(positionalValues.Count);
-
-            for (int i = 0; i < keyProperties.Count; i++)
+            
+            int i = 0; 
+            foreach (IEdmProperty keyProperty in type.Key())

```

Comment:
>type.Key() [](start = 49, length = 10)

why call Key() twice in one method. It's also performance

Reply:
That's a good point @xuzhg - random access is preferable when IList objects boil down to List, since it implements random access as O(1) operation. However, when we are not absolutely sure what's the underlying implementation we should probably do foreach(...) instead of for(...) loop. Moreover, foreach(...) VS for(...) loop is a micro optimization that I don't feel too strong about it either.

To help with this, we should start using the underlying class, instead of the interface (unless when there is a need to send down the interface).